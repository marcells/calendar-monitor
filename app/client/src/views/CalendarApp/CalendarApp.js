import React, { Component } from 'react';
import { connect } from 'react-redux'
import { loadCalendar } from '../../redux/actions';
import Calendar from '../../components/Calendar/Calendar';
import Upcoming from '../../components/Upcoming/Upcoming';
import './CalendarApp.css';

class CalendarApp extends Component {
  async componentDidMount() {
    this.props.dispatch(loadCalendar(this.props.calendar));

    this._timer = setInterval(() => this.props.dispatch(loadCalendar(this.props.calendar)), 10000);
  }

  componentWillUnmount() {
    if (this._timer) {
      clearInterval(this._timer);
    }
  }

  render() {
    return (
      <div className="Content">
        <div className="Calendars">
          { this.props.calendars.map(x => <Calendar key={this.props.calendar + x.date.year + x.date.month} calendar={x} />) }
        </div>

        <Upcoming events={this.props.upcoming} />
      </div>
    );
  }
}

function mapStateToProps(state, ownProps) {
  const { calendars, upcoming } = state.calendar;

  return {
    calendarId: ownProps.calendar,
    calendars,
    upcoming
  };
}

export default connect(mapStateToProps)(CalendarApp);
