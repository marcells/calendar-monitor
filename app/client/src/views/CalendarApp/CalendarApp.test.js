import React from 'react';
import { shallow } from 'enzyme';
import ShallowRenderer from 'react-test-renderer/shallow';
import CalendarApp from './CalendarApp';

describe('CalendarApp', () => {
  describe('auto refresh', () => {
    it('will be disabled when unmounted', () => {
      jest.useFakeTimers();

      const mockedDispatch = jest.fn();
      const wrapper = shallow(<CalendarApp.WrappedComponent dispatch={mockedDispatch} calendarId="test" calendars={[]} upcoming={[]} />);
      wrapper.unmount();

      expect(setInterval).toHaveBeenCalledTimes(1);
      expect(clearInterval).toHaveBeenCalledTimes(1);
    });

    it('will be triggered automatically', () => {
      jest.useFakeTimers();

      const mockedDispatch = jest.fn();
      const wrapper = shallow(<CalendarApp.WrappedComponent dispatch={mockedDispatch} calendarId="test" calendars={[]} upcoming={[]} />);
      jest.runOnlyPendingTimers();
      wrapper.unmount();

      expect(setInterval).toHaveBeenCalledTimes(1);
      expect(mockedDispatch).toHaveBeenCalledTimes(2);
      expect(clearInterval).toHaveBeenCalledTimes(1);
    });
  });

  describe('has calendars', () => {
    it('should render calendar items correctly', () => {
      const calendars = [
        {
          date: {
            year: 2018,
            month: 0
          }
        },
        {
          date: {
            year: 2018,
            month: 1
          }
        }
      ];

      const upcoming = [
        {
          id: 1,
          title: 'Event 1'
        }
      ]

      const mockedDispatch = jest.fn();
      const renderer = new ShallowRenderer();
      const result = renderer.render(<CalendarApp.WrappedComponent dispatch={mockedDispatch} calendarId="test" calendars={calendars} upcoming={upcoming} />);

      expect(result).toMatchSnapshot();
    });
  });

  describe('has no calendars', () => {
    it('should render calendar items correctly', () => {
      const mockedDispatch = jest.fn();
      const renderer = new ShallowRenderer();
      const result = renderer.render(<CalendarApp.WrappedComponent dispatch={mockedDispatch} calendarId="test" calendars={[]} upcoming={[]} />);

      expect(result).toMatchSnapshot();
    });

    it('should render an empty unordered list', () => {
      const mockedDispatch = jest.fn();
      const wrapper = shallow(<CalendarApp.WrappedComponent dispatch={mockedDispatch} calendarId="test" calendars={[]} upcoming={[]} />);

      expect(wrapper.find('Connect(Calendar)')).toHaveLength(0);
      expect(wrapper.find('Connect(Upcoming)')).toHaveLength(1);
      expect(wrapper.find('Connect(EventDetails)')).toHaveLength(1);
    });
  });
});
