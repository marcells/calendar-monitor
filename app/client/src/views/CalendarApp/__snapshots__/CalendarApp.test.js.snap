// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarApp has calendars should render calendar items correctly 1`] = `
<div
  className="Content"
>
  <div
    className="Calendars"
  >
    <Connect(Calendar)
      calendar={
        Object {
          "date": Object {
            "month": 0,
            "year": 2018,
          },
        }
      }
    />
    <Connect(Calendar)
      calendar={
        Object {
          "date": Object {
            "month": 1,
            "year": 2018,
          },
        }
      }
    />
  </div>
  <Connect(Upcoming)
    events={
      Array [
        Object {
          "id": 1,
          "title": "Event 1",
        },
      ]
    }
  />
  <Connect(EventDetails) />
</div>
`;

exports[`CalendarApp has no calendars should render calendar items correctly 1`] = `
<div
  className="Content"
>
  <div
    className="Calendars"
  >
    
  </div>
  <Connect(Upcoming)
    events={Array []}
  />
  <Connect(EventDetails) />
</div>
`;
